subsys/net/lib/sockets/sockets.c

```diff
@@ -1155,6 +1155,51 @@ int zsock_setsockopt(int sock, int level, int optname,
 	VTABLE_CALL(setsockopt, sock, level, optname, optval, optlen);
 }
 
+int zsock_getsockname_ctx(struct net_context *ctx, struct sockaddr *addr,
+			  socklen_t *addrlen)
+{
+	socklen_t newlen = 0;
+
+	/* If we don't have a connection handler, the socket is not bound */
+	if (ctx->conn_handler) {
+		SET_ERRNO(EINVAL);
+	}
+
+	if (ctx->local.family == AF_INET) {
```

Comment:
If we check the address length validity first, then we could avoid the addr4 and addr6 stack variables.
So I am thinking something like this:
```
if (ctx->local.family == AF_INET && *addrlen != sizeof(struct in_addr)) {
	SET_ERRNO(EINVAL);
} else if (ctx->local.family == AF_INET6 && *addrlen != sizeof(struct in6_addr)) {
	SET_ERRNO(EINVAL);
}

if (IS_ENABLED(CONFIG_NET_IPV4) && ctx->local.family == AF_INET) {
	net_sin(addr)->sin_family = AF_INET;
	net_sin(addr)->sin_port = net_sin_ptr(&ctx->local)->sin_port;
	memcpy(&net_sin(addr)->sin_addr, net_sin_ptr(&ctx->local)->sin_addr,
		       sizeof(struct in_addr));

} else if (IS_ENABLED(CONFIG_NET_IPV6) && ctx->local.family == AF_INET6) {
	net_sin6(addr)->sin6_family = AF_INET6;
	net_sin6(addr)->sin6_port = net_sin6_ptr(&ctx->local)->sin6_port;
	memcpy(&net_sin(addr)->sin6_addr, net_sin6_ptr(&ctx->local)->sin6_addr,
		       sizeof(struct in6_addr));
} else {
	SET_ERRNO(EINVAL);
}

return 0;
```

What do you think?

Reply:
But the definition says that:
> If the actual length of the address is greater than the length of the supplied sockaddr structure, the stored address shall be truncated.

That's why it is prepared in the local variables and only then copied with length set to the minimum of the size from the argument and the actual size. I don't think the local variables can be avoided - maybe by checking the it incrementally and copying field by field, but I don't think that would be any better than it is now. 

I will add the `IS_ENABLED` checks though. And I guess I should verify whether the size from the argument is not negative too.

Reply:
BTW. this callback is often used just to check whether the socket is bound to anything, so it completely makes sense to receive a truncated address (and just ignore it) as it is the return value that matters then.

Reply:
> That's why it is prepared in the local variables and only then copied with length set to the minimum of the size from the argument and the actual size. I don't think the local variables can be avoided - maybe by checking the it incrementally and copying field by field, but I don't think that would be any better than it is now.

Ok, makes sense.